File: agent/training/calibration.py
#!/usr/bin/env python3
"""
Model Calibration Module - Platt Scaling Implementation

Calibrates ML model probability outputs to reflect true win rates.
Uses Platt Scaling (logistic regression on model outputs) to transform
raw probabilities into calibrated probabilities.

Problem: Model says 80% confidence, actual win rate might be 60%
Solution: Platt Scaling learns the mapping from raw → calibrated probabilities

ECE (Expected Calibration Error) target: < 0.10
"""

import os
import json
import numpy as np
from datetime import datetime
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass, asdict

# Optional sklearn import for Platt scaling
try:
    from sklearn.linear_model import LogisticRegression
    from sklearn.calibration import calibration_curve
    SKLEARN_AVAILABLE = True
except ImportError:
    SKLEARN_AVAILABLE = False
    print("sklearn not available. Install with: pip install scikit-learn")


@dataclass
class PlattParameters:
    """Platt Scaling parameters for probability calibration."""
    A: float  # Slope parameter
    B: float  # Intercept parameter
    model_name: str
    fitted_at: str
    validation_samples: int
    ece_before_calibration: float
    ece_after_calibration: float
    brier_score_before: float
    brier_score_after: float


@dataclass
class CalibrationEvaluation:
    """Calibration quality metrics."""
    ece: float  # Expected Calibration Error
    mce: float  # Maximum Calibration Error
    brier_score: float
    calibration_curve: List[Dict]
    is_well_calibrated: bool  # ECE < 0.10


class ModelCalibrator:
    """
    Model Calibration using Platt Scaling.
    
    Fits logistic regression on model outputs to calibrate probabilities.
    """
    
    def __init__(self, num_bins: int = 10):
        self.num_bins = num_bins
        self.params: Optional[PlattParameters] = None
        self.calibrator: Optional[LogisticRegression] = None
    
    def fit(
        self,
        predictions: np.ndarray,
        true_labels: np.ndarray,
        model_name: str
    ) -> PlattParameters:
        """
        Fit Platt Scaling on validation data.
        
        Args:
            predictions: Raw model probabilities (0-1)
            true_labels: Actual binary labels (0 or 1)
            model_name: Name of the model for saving
            
        Returns:
            Fitted Platt parameters
        """
        if not SKLEARN_AVAILABLE:
            raise ImportError("sklearn required for calibration")
        
        n = len(predictions)
        if n < 10:
            raise ValueError("Need at least 10 samples for calibration")
        
        print(f"\n📊 Fitting Platt Scaling for {model_name}...")
        print(f"  Samples: {n}")
        print(f"  Positive rate: {np.mean(true_labels):.2%}")
        
        # Evaluate ECE before calibration
        ece_before = self.calculate_ece(predictions, true_labels)
        brier_before = self.calculate_brier_score(predictions, true_labels)
        
        print(f"  ECE before: {ece_before:.4f}")
        print(f"  Brier before: {brier_before:.4f}")
        
        # Fit logistic regression on log-odds
        # Clip predictions to avoid log(0)
        epsilon = 1e-7
        clipped = np.clip(predictions, epsilon, 1 - epsilon)
        
        # Reshape for sklearn
        X = clipped.reshape(-1, 1)
        y = true_labels
        
        # Fit calibrator
        self.calibrator = LogisticRegression(solver='lbfgs', max_iter=1000)
        self.calibrator.fit(X, y)
        
        # Extract Platt parameters (A = coef, B = intercept)
        A = float(self.calibrator.coef_[0][0])
        B = float(self.calibrator.intercept_[0])
        
        # Get calibrated predictions
        calibrated = self.calibrator.predict_proba(X)[:, 1]
        
        # Evaluate ECE after calibration
        ece_after = self.calculate_ece(calibrated, true_labels)
        brier_after = self.calculate_brier_score(calibrated, true_labels)
        
        print(f"  ECE after: {ece_after:.4f}")
        print(f"  Brier after: {brier_after:.4f}")
        print(f"  ECE improvement: {ece_before - ece_after:.4f}")
        
        self.params = PlattParameters(
            A=A,
            B=B,
            model_name=model_name,
            fitted_at=datetime.now().isoformat(),
            validation_samples=n,
            ece_before_calibration=ece_before,
            ece_after_calibration=ece_after,
            brier_score_before=brier_before,
            brier_score_after=brier_after,
        )
        
        return self.params
    
    def calibrate(self, raw_proba: np.ndarray) -> np.ndarray:
        """Apply calibration to raw probabilities."""
        if self.calibrator is None:
            return raw_proba

        epsilon = 1e-7
        clipped = np.clip(raw_proba, epsilon, 1 - epsilon)
        X = clipped.reshape(-1, 1)
        return self.calibrator.predict_proba(X)[:, 1]

    def calculate_ece(
        self,
        predictions: np.ndarray,
        true_labels: np.ndarray
    ) -> float:
        """
        Calculate Expected Calibration Error (ECE).

        ECE measures how well predicted probabilities match actual outcomes.
        Lower is better. Target: < 0.10
        """
        n = len(predictions)
        if n == 0:
            return 0.0

        bins = np.linspace(0, 1, self.num_bins + 1)
        ece = 0.0

        for i in range(self.num_bins):
            mask = (predictions >= bins[i]) & (predictions < bins[i + 1])
            if np.sum(mask) > 0:
                avg_pred = np.mean(predictions[mask])
                avg_actual = np.mean(true_labels[mask])
                ece += (np.sum(mask) / n) * np.abs(avg_pred - avg_actual)

        return float(ece)

    def calculate_brier_score(
        self,
        predictions: np.ndarray,
        true_labels: np.ndarray
    ) -> float:
        """
        Calculate Brier Score.

        Mean squared error between predictions and outcomes.
        Lower is better. Range: 0-1
        """
        return float(np.mean((predictions - true_labels) ** 2))

    def evaluate(
        self,
        predictions: np.ndarray,
        true_labels: np.ndarray
    ) -> CalibrationEvaluation:
        """Evaluate calibration quality."""
        ece = self.calculate_ece(predictions, true_labels)
        brier = self.calculate_brier_score(predictions, true_labels)

        # Calculate calibration curve
        bins = np.linspace(0, 1, self.num_bins + 1)
        curve = []
        mce = 0.0

        for i in range(self.num_bins):
            mask = (predictions >= bins[i]) & (predictions < bins[i + 1])
            if np.sum(mask) > 0:
                avg_pred = np.mean(predictions[mask])
                avg_actual = np.mean(true_labels[mask])
                error = np.abs(avg_pred - avg_actual)
                mce = max(mce, error)

                curve.append({
                    "bin_start": float(bins[i]),
                    "bin_end": float(bins[i + 1]),
                    "mean_predicted": float(avg_pred),
                    "mean_actual": float(avg_actual),
                    "count": int(np.sum(mask)),
                })

        return CalibrationEvaluation(
            ece=ece,
            mce=mce,
            brier_score=brier,
            calibration_curve=curve,
            is_well_calibrated=ece < 0.10,
        )

    def save(self, output_dir: str) -> str:
        """Save calibration parameters to JSON file."""
        if self.params is None:
            raise ValueError("No calibration fitted. Call fit() first.")

        os.makedirs(output_dir, exist_ok=True)
        filepath = os.path.join(output_dir, f"{self.params.model_name}_calibration.json")

        with open(filepath, "w") as f:
            json.dump(asdict(self.params), f, indent=2)

        print(f"  Calibration saved to {filepath}")
        return filepath

    @classmethod
    def load(cls, filepath: str) -> "ModelCalibrator":
        """Load calibration parameters from JSON file."""
        with open(filepath, "r") as f:
            data = json.load(f)

        calibrator = cls()
        calibrator.params = PlattParameters(**data)

        # Reconstruct sklearn calibrator from parameters
        if SKLEARN_AVAILABLE:
            calibrator.calibrator = LogisticRegression()
            calibrator.calibrator.coef_ = np.array([[calibrator.params.A]])
            calibrator.calibrator.intercept_ = np.array([calibrator.params.B])
            calibrator.calibrator.classes_ = np.array([0, 1])

        return calibrator


def fit_and_save_calibration(
    predictions: np.ndarray,
    true_labels: np.ndarray,
    model_name: str,
    output_dir: str,
    num_bins: int = 10
) -> PlattParameters:
    """
    Convenience function to fit and save calibration in one step.

    Args:
        predictions: Raw model probabilities
        true_labels: Actual binary labels
        model_name: Name of the model
        output_dir: Directory to save calibration
        num_bins: Number of bins for ECE calculation

    Returns:
        Fitted Platt parameters
    """
    calibrator = ModelCalibrator(num_bins=num_bins)
    params = calibrator.fit(predictions, true_labels, model_name)
    calibrator.save(output_dir)
    return params

File: agent/eliza/src/services/ml/calibration.ts
/**
 * Model Calibration Service - Platt Scaling Implementation
 * 
 * Calibrates ML model probability outputs to reflect true win rates.
 * Uses Platt Scaling (logistic regression on model outputs) to transform
 * raw probabilities into calibrated probabilities.
 * 
 * Problem: Model says 80% confidence, actual win rate might be 60%
 * Solution: Platt Scaling learns the mapping from raw → calibrated probabilities
 * 
 * ECE (Expected Calibration Error) target: < 0.10
 */

import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { logger } from '../logger.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Calibration parameters directory
const CALIBRATION_DIR = path.join(__dirname, '../../../models/calibration');

/**
 * Platt Scaling Parameters
 * Transforms raw probability p to calibrated probability:
 * calibrated = 1 / (1 + exp(A * p + B))
 */
export interface PlattParameters {
  A: number;  // Slope parameter
  B: number;  // Intercept parameter
  modelName: string;
  fittedAt: string;
  validationSamples: number;
  eceBeforeCalibration: number;
  eceAfterCalibration: number;
  brierScoreBefore: number;
  brierScoreAfter: number;
}

/**
 * Calibration configuration
 */
export interface CalibrationConfig {
  enabled: boolean;
  method: 'platt' | 'isotonic';
  minEceImprovement: number;
  saveCalibrationCurves: boolean;
  numBins: number;  // For ECE calculation
}

/**
 * Calibration curve data point
 */
export interface CalibrationBin {
  binStart: number;
  binEnd: number;
  meanPredicted: number;
  meanActual: number;
  count: number;
}

/**
 * Calibration evaluation result
 */
export interface CalibrationEvaluation {
  ece: number;  // Expected Calibration Error
  mce: number;  // Maximum Calibration Error
  brierScore: number;
  calibrationCurve: CalibrationBin[];
  isWellCalibrated: boolean;  // ECE < 0.10
}

/**
 * Model Calibration Service
 * 
 * Singleton service for calibrating ML model outputs using Platt Scaling.
 */
export class CalibrationService {
  private static instance: CalibrationService | null = null;
  private calibrationParams: Map<string, PlattParameters> = new Map();
  private config: CalibrationConfig;
  private initialized = false;

  private constructor(config?: Partial<CalibrationConfig>) {
    this.config = {
      enabled: true,
      method: 'platt',
      minEceImprovement: 0.02,
      saveCalibrationCurves: true,
      numBins: 10,
      ...config,
    };
  }

  static getInstance(config?: Partial<CalibrationConfig>): CalibrationService {
    if (!CalibrationService.instance) {
      CalibrationService.instance = new CalibrationService(config);
    }
    return CalibrationService.instance;
  }

  static resetInstance(): void {
    CalibrationService.instance = null;
  }

  /**
   * Initialize calibration service and load saved parameters
   */
  async initialize(): Promise<boolean> {
    if (this.initialized) return true;

    try {
      // Ensure calibration directory exists
      if (!fs.existsSync(CALIBRATION_DIR)) {
        fs.mkdirSync(CALIBRATION_DIR, { recursive: true });
        logger.info('[Calibration] Created calibration directory', { path: CALIBRATION_DIR });
      }

      // Load existing calibration parameters
      const files = fs.readdirSync(CALIBRATION_DIR);
      for (const file of files) {
        if (file.endsWith('_calibration.json')) {
          const modelName = file.replace('_calibration.json', '');
          await this.loadCalibration(modelName);
        }
      }

      this.initialized = true;
      logger.info('[Calibration] Service initialized', {
        loadedModels: Array.from(this.calibrationParams.keys()),
      });
      return true;
    } catch (error) {
      logger.error('[Calibration] Failed to initialize', { error });
      return false;
    }
  }

  isInitialized(): boolean {
    return this.initialized;
  }

  isEnabled(): boolean {
    return this.config.enabled;
  }

  hasCalibration(modelName: string): boolean {
    return this.calibrationParams.has(modelName);
  }

  /**
   * Fit Platt Scaling parameters on validation data
   *
   * Uses Newton-Raphson optimization to find A and B parameters
   * that minimize negative log-likelihood.
   *
   * @param predictions Raw model probabilities (0-1)
   * @param trueLabels Actual binary labels (0 or 1)
   * @param modelName Name of the model for saving
   * @returns Fitted Platt parameters
   */
  fitPlattScaling(
    predictions: number[],
    trueLabels: number[],
    modelName: string
  ): PlattParameters {
    if (predictions.length !== trueLabels.length) {
      throw new Error('Predictions and labels must have same length');
    }

    const n = predictions.length;
    if (n < 10) {
      throw new Error('Need at least 10 samples for calibration');
    }

    logger.info('[Calibration] Fitting Platt Scaling', {
      modelName,
      samples: n,
      positiveRate: trueLabels.filter(l => l === 1).length / n,
    });

    // Evaluate ECE before calibration
    const eceBefore = this.calculateECE(predictions, trueLabels);
    const brierBefore = this.calculateBrierScore(predictions, trueLabels);

    // Platt Scaling: fit logistic regression on log-odds
    // Transform predictions to log-odds (avoid 0 and 1)
    const epsilon = 1e-7;
    const clippedPreds = predictions.map(p => Math.max(epsilon, Math.min(1 - epsilon, p)));

    // Newton-Raphson optimization for Platt parameters
    // Initialize A and B
    let A = 0;
    let B = 0;
    const maxIter = 100;
    const tolerance = 1e-8;

    // Target values with Laplace smoothing (Platt's original paper)
    const nPos = trueLabels.filter(l => l === 1).length;
    const nNeg = n - nPos;
    const targetPos = (nPos + 1) / (nPos + 2);
    const targetNeg = 1 / (nNeg + 2);
    const targets = trueLabels.map(l => l === 1 ? targetPos : targetNeg);

    for (let iter = 0; iter < maxIter; iter++) {
      // Compute calibrated probabilities
      const calibrated = clippedPreds.map(p => 1 / (1 + Math.exp(A * p + B)));

      // Compute gradient and Hessian
      let gradA = 0, gradB = 0;
      let hessAA = 0, hessAB = 0, hessBB = 0;

      for (let i = 0; i < n; i++) {
        const p = clippedPreds[i];
        const q = calibrated[i];
        const t = targets[i];
        const d = q * (1 - q);

        gradA += p * (q - t);
        gradB += (q - t);
        hessAA += p * p * d;
        hessAB += p * d;
        hessBB += d;
      }

      // Regularization to avoid singular Hessian
      hessAA += 1e-6;
      hessBB += 1e-6;

      // Solve 2x2 system: H * delta = -grad
      const det = hessAA * hessBB - hessAB * hessAB;
      if (Math.abs(det) < 1e-10) break;

      const deltaA = -(hessBB * gradA - hessAB * gradB) / det;
      const deltaB = -(hessAA * gradB - hessAB * gradA) / det;

      A += deltaA;
      B += deltaB;

      if (Math.abs(deltaA) < tolerance && Math.abs(deltaB) < tolerance) {
        break;
      }
    }

    // Evaluate ECE after calibration
    const calibratedPreds = clippedPreds.map(p => 1 / (1 + Math.exp(A * p + B)));
    const eceAfter = this.calculateECE(calibratedPreds, trueLabels);
    const brierAfter = this.calculateBrierScore(calibratedPreds, trueLabels);

    const params: PlattParameters = {
      A,
      B,
      modelName,
      fittedAt: new Date().toISOString(),
      validationSamples: n,
      eceBeforeCalibration: eceBefore,
      eceAfterCalibration: eceAfter,
      brierScoreBefore: brierBefore,
      brierScoreAfter: brierAfter,
    };

    // Store in memory
    this.calibrationParams.set(modelName, params);

    logger.info('[Calibration] Platt Scaling fitted', {
      modelName,
      A: A.toFixed(4),
      B: B.toFixed(4),
      eceBefore: eceBefore.toFixed(4),
      eceAfter: eceAfter.toFixed(4),
      improvement: (eceBefore - eceAfter).toFixed(4),
    });

    return params;
  }

  /**
   * Apply calibration to raw probability
   */
  calibrateProba(rawProba: number, modelName: string): number {
    if (!this.config.enabled) {
      return rawProba;
    }

    const params = this.calibrationParams.get(modelName);
    if (!params) {
      return rawProba;
    }

    // Clip to avoid numerical issues
    const epsilon = 1e-7;
    const clipped = Math.max(epsilon, Math.min(1 - epsilon, rawProba));

    // Apply Platt transformation
    const calibrated = 1 / (1 + Math.exp(params.A * clipped + params.B));

    return calibrated;
  }

  /**
   * Calibrate array of probabilities
   */
  calibrateProbas(rawProbas: number[], modelName: string): number[] {
    return rawProbas.map(p => this.calibrateProba(p, modelName));
  }

  /**
   * Calculate Expected Calibration Error (ECE)
   *
   * ECE measures how well predicted probabilities match actual outcomes.
   * Lower is better. Target: < 0.10
   */
  calculateECE(predictions: number[], trueLabels: number[]): number {
    const n = predictions.length;
    if (n === 0) return 0;

    const numBins = this.config.numBins;
    const bins: { sum: number; count: number; correct: number }[] =
      Array.from({ length: numBins }, () => ({ sum: 0, count: 0, correct: 0 }));

    for (let i = 0; i < n; i++) {
      const binIdx = Math.min(Math.floor(predictions[i] * numBins), numBins - 1);
      bins[binIdx].sum += predictions[i];
      bins[binIdx].count += 1;
      bins[binIdx].correct += trueLabels[i];
    }

    let ece = 0;
    for (const bin of bins) {
      if (bin.count > 0) {
        const avgPred = bin.sum / bin.count;
        const avgActual = bin.correct / bin.count;
        ece += (bin.count / n) * Math.abs(avgPred - avgActual);
      }
    }

    return ece;
  }

  /**
   * Calculate Brier Score
   *
   * Mean squared error between predictions and outcomes.
   * Lower is better. Range: 0-1
   */
  calculateBrierScore(predictions: number[], trueLabels: number[]): number {
    const n = predictions.length;
    if (n === 0) return 0;

    let sum = 0;
    for (let i = 0; i < n; i++) {
      sum += Math.pow(predictions[i] - trueLabels[i], 2);
    }

    return sum / n;
  }

  /**
   * Evaluate calibration quality
   */
  evaluateCalibration(
    predictions: number[],
    trueLabels: number[]
  ): CalibrationEvaluation {
    const ece = this.calculateECE(predictions, trueLabels);
    const brierScore = this.calculateBrierScore(predictions, trueLabels);

    // Calculate calibration curve
    const numBins = this.config.numBins;
    const calibrationCurve: CalibrationBin[] = [];

    for (let i = 0; i < numBins; i++) {
      const binStart = i / numBins;
      const binEnd = (i + 1) / numBins;

      const inBin = predictions
        .map((p, idx) => ({ p, label: trueLabels[idx] }))
        .filter(({ p }) => p >= binStart && p < binEnd);

      if (inBin.length > 0) {
        const meanPredicted = inBin.reduce((s, { p }) => s + p, 0) / inBin.length;
        const meanActual = inBin.reduce((s, { label }) => s + label, 0) / inBin.length;

        calibrationCurve.push({
          binStart,
          binEnd,
          meanPredicted,
          meanActual,
          count: inBin.length,
        });
      }
    }

    // Calculate MCE (Maximum Calibration Error)
    let mce = 0;
    for (const bin of calibrationCurve) {
      const error = Math.abs(bin.meanPredicted - bin.meanActual);
      mce = Math.max(mce, error);
    }

    return {
      ece,
      mce,
      brierScore,
      calibrationCurve,
      isWellCalibrated: ece < 0.10,
    };
  }

  /**
   * Save calibration parameters to file
   */
  async saveCalibration(modelName: string): Promise<void> {
    const params = this.calibrationParams.get(modelName);
    if (!params) {
      throw new Error(`No calibration found for model: ${modelName}`);
    }

    const filePath = path.join(CALIBRATION_DIR, `${modelName}_calibration.json`);
    fs.writeFileSync(filePath, JSON.stringify(params, null, 2));

    logger.info('[Calibration] Saved calibration', { modelName, path: filePath });
  }

  /**
   * Load calibration parameters from file
   */
  async loadCalibration(modelName: string): Promise<PlattParameters | null> {
    const filePath = path.join(CALIBRATION_DIR, `${modelName}_calibration.json`);

    if (!fs.existsSync(filePath)) {
      logger.debug('[Calibration] No calibration file found', { modelName });
      return null;
    }

    try {
      const data = fs.readFileSync(filePath, 'utf-8');
      const rawParams = JSON.parse(data);

      // Handle both snake_case (Python) and camelCase (TypeScript) field names
      const params: PlattParameters = {
        A: rawParams.A,
        B: rawParams.B,
        modelName: rawParams.modelName || rawParams.model_name || modelName,
        fittedAt: rawParams.fittedAt || rawParams.fitted_at || new Date().toISOString(),
        validationSamples: rawParams.validationSamples || rawParams.validation_samples || 0,
        eceBeforeCalibration: rawParams.eceBeforeCalibration || rawParams.ece_before_calibration || 0,
        eceAfterCalibration: rawParams.eceAfterCalibration || rawParams.ece_after_calibration || 0,
        brierScoreBefore: rawParams.brierScoreBefore || rawParams.brier_score_before || 0,
        brierScoreAfter: rawParams.brierScoreAfter || rawParams.brier_score_after || 0,
      };

      this.calibrationParams.set(modelName, params);

      logger.info('[Calibration] Loaded calibration', {
        modelName,
        A: params.A.toFixed(4),
        B: params.B.toFixed(4),
        ece: params.eceAfterCalibration.toFixed(4),
      });

      return params;
    } catch (error) {
      logger.error('[Calibration] Failed to load calibration', { modelName, error });
      return null;
    }
  }

  /**
   * Get calibration parameters for a model
   */
  getCalibrationParams(modelName: string): PlattParameters | undefined {
    return this.calibrationParams.get(modelName);
  }

  /**
   * Get all loaded calibrations
   */
  getAllCalibrations(): Map<string, PlattParameters> {
    return new Map(this.calibrationParams);
  }
}

// Export singleton getter
export function getCalibrationService(config?: Partial<CalibrationConfig>): CalibrationService {
  return CalibrationService.getInstance(config);
}

// Export singleton instance
export const calibrationService = CalibrationService.getInstance();
